ui-v2/tests/steps.js

```diff
@@ -0,0 +1,232 @@
+        return create(number, model, value);
+      })
+      .given(
+        ['$number $model model[s]? from yaml\n$yaml', '$number $model model from json\n$json'],
+        function(number, model, data) {
+          return create(number, model, data);
+        }
+      )
+      // interactions
+      .when('I visit the $name page', function(name) {
+        currentPage = pages[name];
+        return currentPage.visit();
+      })
+      .when('I visit the $name page for the "$id" $model', function(name, id, model) {
+        currentPage = pages[name];
+        return currentPage.visit({
+          [model]: id,
+        });
+      })
+      .when(
+        ['I visit the $name page for yaml\n$yaml', 'I visit the $name page for json\n$json'],
+        function(name, data) {
+          currentPage = pages[name];
+          return pages[name].visit(data);
+        }
+      )
+      .when('I click "$selector"', function(selector) {
+        return click(selector);
+      })
+      .when('I click $prop on the $component', function(prop, component) {
+        // Collection
+        var obj;
+        if (typeof currentPage[component].objectAt === 'function') {
+          obj = currentPage[component].objectAt(0);
+        } else {
+          obj = currentPage[component];
+        }
+        const func = obj[prop].bind(obj);
+        try {
+          return func();
+        } catch (e) {
+          console.error(e);
+          throw new Error(`The '${prop}' property on the '${component}' page object doesn't exist`);
+        }
+      })
+      .when('I submit', function(selector) {
+        return currentPage.submit();
+      })
+      .then('I fill in "$name" with "$value"', function(name, value) {
+        return currentPage.fillIn(name, value);
+      })
+      .then(['I fill in with yaml\n$yaml', 'I fill in with json\n$json'], function(data) {
+        return Object.keys(data).reduce(function(prev, item, i, arr) {
+          return prev.fillIn(item, data[item]);
+        }, currentPage);
+      })
+      .then(['I type with yaml\n$yaml'], function(data) {
+        const keys = Object.keys(data);
+        return keys
+          .reduce(function(prev, item, i, arr) {
+            return prev.fillIn(item, data[item]);
+          }, currentPage)
+          .then(function() {
+            return Promise.all(
+              keys.map(function(item) {
+                return triggerKeyEvent(`[name="${item}"]`, 'keyup', 83);
+              })
+            );
+          });
+      })
+      // debugging helpers
+      .then('print the current url', function(url) {
+        console.log(currentURL());
+        return Promise.resolve();
+      })
+      .then('log the "$text"', function(text) {
+        console.log(text);
+        return Promise.resolve();
+      })
+      .then('pause for $milliseconds', function(milliseconds) {
+        return new Promise(function(resolve) {
+          setTimeout(resolve, milliseconds);
```

Comment:
It would be better to use the runloop's timers instead of vanilla setTimeout, (i.e., `run.later(resolve, milliseconds)`)

Reply:
Happy to look at changing it. Out of interest, why do you think it would be better?

Reply:
Any asynchronous code that isn't threaded through the runloop can lead to weird race conditions or the classic error `Assertion failed: You have turned on testing mode, which disabled the run-loop’s autorun. You will need to wrap any code with asynchronous side-effects in an Ember.run`. Granted that error will only happen when your implementation has asynchronous code outside the run loop, but I find it's easier to use the runloop for everything rather than only when necessary.

Reply:
Hmm need to think about this one some more. My current thinking is that I'm trying to reproduce actual user interaction as much as possible here - outside of ember itself, that way its as close to 'acceptance' or e2e as possible. Therefore I'd rather use a non-ember construct to perform the equivalent of the user pausing to look at a page.

Actually, just realizing now, I'm only using this for debugging purposes to I can get a look at the rendered page myself - so no biggie whichever. Would using ember specific stuff here possibly prevent a page from being rendered? I doubt it but the above thing about replicating user interaction as closely as possible stands here.